WIP: Windows Build

jmittert opened the pull request:
Enable compilation of a Windows target. This doesn't run yet, the stubs
all needed to be filled in, but compilation completes successfully with
the areas stubbed out.

This is a work in progress so far, I was just hoping for some feedback or thoughts in the mean time so I'm not dumping a bunch of code out of nowhere down the road.

jakepetroules commented on include/llbuild/Basic/CrossPlatformCompatibility.h:
```diff
@@ -18,16 +18,47 @@
 #define CrossPlatformCompatibility_h
 
 #ifdef _MSC_VER
```
Assuming you're using MSVC and not Clang for testing thing, I'd just like to note for the record that Clang defines both _MSC_VER and _WIN32 for compatibility with MSVC when targeting Windows, so this branch actually does mean Windows in general, not *only* Windows+MSVC.

If you definitely need something to be MSVC specific and NOT apply to Clang, you'll need to use something like `#if (defined(_WIN32) || defined(_MSC_VER)) && !defined(__clang__)`

jakepetroules commented on include/llbuild/Basic/CrossPlatformCompatibility.h:
```diff
@@ -18,16 +18,47 @@
+#endif
+
+#if defined(_WIN32)
+using llbuild_pid_t = HANDLE;
```
Why `using` instead of just using the typedef we already defined above?

dmbryson commented on lib/BuildSystem/BuildSystem.cpp:
```diff
@@ -1009,7 +1013,11 @@ class FilteredDirectoryContentsTask : public Task {
       bool excluded = false;
       for (auto pattern : filterStrings) {
+#if defined(_WIN32)
+        if (PathMatchSpecA(filename.c_str(), pattern.data())) {
```
Might be worth putting this into a method in PlatformUtility.   Would be nice to try to keep the _WIN32 ifdefs contained to that layer as much as possible.

jakepetroules replied:
PathMatchSpecW

dmbryson commented on products/libllbuild/include/llbuild/buildsystem.h:
```diff
@@ -24,18 +24,7 @@
-
-typedef pid_t llbuild_pid_t;
-#endif
+#include "llbuild/Basic/CrossPlatformCompatibility.h"
```
These are here because the C library headers are distinct from the C++ headers and thus should not depend on that header being available.

dmbryson commented on unittests/CAPI/BuildSystem-C-API.cpp:
```diff
@@ -340,7 +347,11 @@ TEST(BuildSystemCAPI, CustomToolWithDiscoveredDependencies) {
   // times on the granularity of a second.
   // FIXME: might be better to just fake the timestamp on the file
+#if defined(_WIN32)
+  Sleep(1000);
```
Another method we should probably put in PlatformUtility

jakepetroules commented on include/llbuild/Basic/CrossPlatformCompatibility.h:
```diff
@@ -18,16 +18,47 @@
 
 typedef pid_t llbuild_pid_t;
+typedef int FD;
+#endif
```
Can you please add `// #ifdef _MSC_VER` to the end of this line? I misread this section like 5 times :S

jakepetroules commented on include/llbuild/Basic/PlatformUtility.h:
```diff
@@ -41,7 +43,11 @@ int write(int fileHandle, void *destinationBuffer, unsigned int maxCharCount);
+#if defined(_WIN32)
+int raiseOpenFileLimit(int limit = 2048);
+#else
 int raiseOpenFileLimit(rlim_t limit = 2048);
```
Probably could use another typedef to minimize ifdefs

jakepetroules commented on lib/Basic/PlatformUtility.cpp:
```diff
@@ -136,6 +136,16 @@ int sys::write(int fileHandle, void *destinationBuffer,
 #endif
 }
 
+#if defined(_WIN32)
```
In conjunction with the typedef given above, please minimize the scope of the ifdefs to be inside the function definition

jakepetroules commented on lib/Basic/Subprocess.cpp:
```diff
@@ -207,18 +219,32 @@ class ControlProtocolState {
     char buf[4096];
+#if defined(_WIN32)
+    DWORD numBytes;
+    if (!ReadFile(outputPipe, buf, sizeof(buf), &numBytes, nullptr)){
```
space between ) and {

jakepetroules commented on lib/Basic/Subprocess.cpp:
```diff
@@ -207,18 +219,32 @@ class ControlProtocolState {
+    DWORD numBytes;
+    if (!ReadFile(outputPipe, buf, sizeof(buf), &numBytes, nullptr)){
+      int err = GetLastError();
+      FormatMessage(FORMAT_MESSAGE_FROM_SYSTEM,
```
Please use FormatMessageW in order to support Unicode.

Also, please define `UNICODE` and `_UNICODE` somewhere centrally, to avoid any risk of accidentally using the ASCII functions.

jakepetroules commented on lib/Basic/Subprocess.cpp:
```diff
@@ -243,8 +268,55 @@ static void cleanUpExecutedProcess(
+  if (exitCode != 0 || waitResult == WAIT_FAILED || waitResult == WAIT_ABANDONED) {
+    char buf[4096];
+    auto result = ProcessResult::makeFailed(exitCode);
+    FormatMessage(FORMAT_MESSAGE_FROM_SYSTEM,
```
FormatMessageW

jakepetroules commented on lib/Basic/Subprocess.cpp:
```diff
@@ -319,6 +390,27 @@ void llbuild::basic::spawnProcess(
+
+
+#if defined(_WIN32)
+  //TODO: Implement using CreateProcessW on Windows
```
Please also add an `abort()` for good measure.

jakepetroules commented on lib/Commands/BuildSystemCommand.cpp:
```diff
@@ -380,6 +384,9 @@ static void parseUsage(int exitCode) {
 
 static int executeParseCommand(std::vector<std::string> args) {
+#if defined(_WIN32)
+//TODO jmittert
```
Please change to:

    // TODO: Not yet implemented
    abort();

jakepetroules commented on lib/Commands/BuildSystemCommand.cpp:
```diff
@@ -414,13 +421,17 @@ static int executeParseCommand(std::vector<std::string> args) {
 
 class BasicBuildSystemFrontendDelegate : public BuildSystemFrontendDelegate {
+#if defined(_WIN32)
+//TODO jmittert
```
Please change to:

    // TODO: Not yet implemented
    abort();

jakepetroules commented on lib/Commands/BuildSystemCommand.cpp:
```diff
@@ -523,10 +534,15 @@ class BasicBuildSystemFrontendDelegate : public BuildSystemFrontendDelegate {
 };
 
+#if defined(_WIN32)
+//TODO jmittert
```
Please change to:

    // TODO: Not yet implemented
    abort();

jakepetroules commented on lib/Commands/BuildSystemCommand.cpp:
```diff
@@ -538,6 +554,9 @@ static void buildUsage(int exitCode) {
 
 static int executeBuildCommand(std::vector<std::string> args) {
+#if defined(_WIN32)
+//TODO jmittert
```
Please change to:

    // TODO: Not yet implemented
    abort();

jakepetroules commented on lib/Commands/NinjaBuildCommand.cpp:
```diff
@@ -413,7 +417,11 @@ class BuildContext : public ExecutionQueueDelegate {
 
   /// The previous SIGINT handler.
+#if defined(_WIN32)
+//TODO jmittert
```
Please change to:

    // TODO: Not yet implemented
    abort();

jakepetroules commented on lib/Commands/NinjaBuildCommand.cpp:
```diff
@@ -485,6 +493,9 @@ class BuildContext : public ExecutionQueueDelegate {
       isCancelled(false)
   {
+#if defined(_WIN32)
+//TODO jmittert
```
Please change to:

    // TODO: Not yet implemented
    abort();

jakepetroules commented on lib/Commands/NinjaBuildCommand.cpp:
```diff
@@ -508,9 +519,13 @@ class BuildContext : public ExecutionQueueDelegate {
 
   ~BuildContext() {
+#if defined(_WIN32)
+//TODO jmittert
```
Please change to:

    // TODO: Not yet implemented
    abort();

jakepetroules commented on lib/Commands/NinjaBuildCommand.cpp:
```diff
@@ -1509,6 +1525,9 @@ void NinjaBuildEngineDelegate::error(const Twine& message) {
 
 int commands::executeNinjaBuildCommand(std::vector<std::string> args) {
+#if defined(_WIN32)
+//TODO jmittert
```
Please change to:

    // TODO: Not yet implemented
    abort();

jakepetroules commented on unittests/CAPI/BuildSystem-C-API.cpp:
```diff
@@ -59,6 +59,9 @@ depinfo_tester_command_execute_command(void *context,
                                        llb_task_t* task,
                                        llb_buildsystem_queue_job_context_t* job) {
+#if defined(_WIN32)
+//TODO jmittert
```
Please change to:

    // TODO: Not yet implemented
    abort();

jakepetroules commented on unittests/CAPI/BuildSystem-C-API.cpp:
```diff
@@ -255,6 +259,9 @@ static void command_process_finished(void* context,
 
 TEST(BuildSystemCAPI, CustomToolWithDiscoveredDependencies) {
+#if defined(_WIN32)
+//TODO jmittert
```
Please change to:

    // TODO: Not yet implemented
    abort();

jmittert commented:
Alright:
- Moved the ifdef functions in to platformcompat
- Used W functions on Windows
- Changed the todos to not yet implemented
- Ran through clang format

For going in and implementing, would you prefer more commits in this PR, or separate PRs for each smaller change?

jakepetroules commented on lib/Basic/CMakeLists.txt:
```diff
@@ -15,3 +15,8 @@ add_llbuild_library(llbuildBasic
 endif()
+
+if (${CMAKE_SYSTEM_NAME} STREQUAL "Windows")
+  link_directories(${LLBUILD_WINDOWS_UM_DIR})
```
Why is CMake not able to figure this out on its own? That directory should already be in the search path when building from a VS command prompt AFAIK.

jakepetroules commented on include/llbuild/Basic/PlatformUtility.h:
```diff
@@ -41,8 +46,17 @@ int write(int fileHandle, void *destinationBuffer, unsigned int maxCharCount);
+// Returns MATCH if a match is detected, NO_MATCH if there is no match, and
+// MATCH_ERROR on error. Windows callers may use GetLastError to get additional
+// error information.
+MATCH_RESULT filenameMatch(std::string pattern, std::string filename);
```
pass the std::strings with const&?

dmbryson commented:
@swift-ci please smoke test

dmbryson commented:
I think separate PRs for smaller changes would be a fine way to go.

jakepetroules commented:
@swift-ci please smoke test

jakepetroules commented:
@swift-ci please smoke test
